### ISSUE TITLE
`CollectionAssociation#size` should use `@association_ids` if available

### ISSUE THREAD

**Issue by chancancode** [rails/rails#32569]
### Steps to reproduce

```ruby
# frozen_string_literal: true

begin
  require "bundler/inline"
rescue LoadError => e
  $stderr.puts "Bundler version 1.10 or later is required. Please update your Bundler"
  raise e
end

gemfile(true) do
  source "https://rubygems.org"

  git_source(:github) { |repo| "https://github.com/#{repo}.git" }

  gem "rails", github: "rails/rails"
  gem "sqlite3"
end

require "active_record"
require "minitest/autorun"
require "logger"

# This connection will do for database-independent bug reports.
ActiveRecord::Base.establish_connection(adapter: "sqlite3", database: ":memory:")
ActiveRecord::Base.logger = Logger.new(STDOUT)

ActiveRecord::Schema.define do
  create_table :users, force: true do |t|
  end

  create_table :posts, force: true do |t|
    t.integer :user_id
  end
end

class User < ActiveRecord::Base
  has_many :posts
end

class Post < ActiveRecord::Base
  belongs_to :user
end

class BugTest < Minitest::Test
  def test_association_size
    user = User.create!
    100.times { Post.create(user_id: user.id) }

    # This works: it loads the association IDs
    assert_sql_query(/SELECT "posts".id FROM "posts"/) do
      user.post_ids
    end

    # This works: once loaded, it's cached (until reset/reload)
    assert_no_sql_queries do
      assert_equal 100, user.post_ids.length
    end

    # I think this should work, but it does (it still goes to the database for a `SELECT COUNT`)
    assert_no_sql_queries do
      assert_equal 100, user.posts.size
    end
  end

  private

  def assert_sql_query(query)
    # implement me, or just look at the logs
    yield
  end

  def assert_no_sql_queries
    # implement me, or just look at the logs
    yield
  end
end
```

### Expected behavior

If `user.post_ids` is already loaded and cached, I think `user.posts.size` should not go to the DB.

### Actual behavior

Even if `user.post_ids` is already loaded and cached, `user.posts.size` still goes to the DB to do a `SELECT COUNT`.

**Comment by chancancode** [comment]
The relevant code is [here](https://github.com/rails/rails/blob/master/activerecord/lib/active_record/associations/collection_association.rb#L213-L214). IMO, it should do something like this:

```ruby
      def size
        if !find_target? || has_known_size?
          loaded? ? target.size : @association_ids.length
        elsif !association_scope.group_values.empty?
          load_target.size
        elsif !association_scope.distinct_value && target.is_a?(Array)
          unsaved_records = target.select(&:new_record?)
          unsaved_records.size + count_records
        else
          count_records
        end
      end

      def empty?
        if has_known_size?
          size.zero?
        else
          @target.blank? && !scope.exists?
        end
      end

      private

        def has_known_size?
          loaded? || @association_ids
        end
```

**Comment by chancancode** [comment]
https://github.com/TheOdinProject/theodinproject/pull/890 is a (slightly more involved) real world example of this.

**Comment by pixeltrix** [comment]
There's an edge condition where `size` will not go to the database after this change and you might have stale information now so dubious whether we should backport it. However more importantly there's a slight issue with the diff as presented because if `!find_target?` is true because the parent is a new record that hasn't been saved yet then it will try to look up the length from `@association_ids` which is nil. Other than that logic bug I reckon it's a 👍 change.

**Comment by chancancode** [comment]
Sounds good to me. I agree it’s not obvious that we should backport. Also thanks for catching the bug.

For the edge case, I think that’s no worse today, since you can already do a size (which does a count and caches it in the query cache), then another thread can update it so that could result in `collection.size != collection.length`.

If anyone knows where to look for the existing tests for `#size` please do let us know in the comments. I figured that might be the hardest part if someone were to work on this as their first patch.

**Comment by chancancode** [comment]
@pixeltrix I updated the diff to do a slightly less fancy thing for now, so we can focus on the behavior. Do you happen to know why `#empty?` doesn't check the `!find_target?` case you talked about?

**Comment by pixeltrix** [comment]
@chancancode looking at the history I'd say that since the implementation of `empty?` when `!find_target?` was added was `size.zero?` it obviously didn't need it. When `empty?` became more complex it's probably been overlooked.

**Comment by pixeltrix** [comment]
BTW, tests for `size` seem to be distributed over a few files in Active Record - I'd probably add any for this feature to `has_many_association_test.rb`

**Comment by juliusdelta** [comment]
@pixeltrix I'd like to take this if it's unassigned. First non-documentation related patch for me.

**Comment by tgturner** [comment]
I am also interested in this issue!

**Comment by juliusdelta** [comment]
I assumed I was to wait to be assigned since good first patches are few and far between. @pixeltrix Is this etiquette? Or should I just open the PR next time.

**Comment by tgturner** [comment]
@juliusdelta I apologize! Guess I got a little trigger happy. I assumed that you had decided not to work on it since it was small and had been a day, didn’t realize you were waiting for it to be assigned. I’m happy to close the PR and let you pick this one up if you’d like.

**Comment by juliusdelta** [comment]
@tgturner No worries! Keep it open for sure. I'm glad you were able to do it. I hope I didn't come off as passive aggressive lol, I'm just genuinely wondering if there is like etiquette or something as far as taking issues goes cause I actually don't know lol.

**Comment by tgturner** [comment]
I'm not sure the etiquette either, but thanks for being so understanding!

**Comment by pixeltrix** [comment]
@juliusdelta @tgturner we can't assign issues in the GitHub UI to people who aren't part of the org AFAIK plus often people might say they're working on an issue but life understandably gets in the way. So the general rule is that whoever files the first PR is accepted if it's a very straightforward issue but for more complicated PRs then we'll generally go with the team's preferred implementation. The competing PR's respective authors can usually agree which is the best approach without the team having to make any intervention.

TL;DR you don't need to ask permission to start work on an issue and if someone has commented that they're working on an issue then give them a few days to get a PR ready. After that feel free to work on it yourself - you can always post a message to ask if the person is still working on it.

**Comment by chancancode** [comment]
Closed by #32617
